keps/sig-cluster-lifecycle/addons/20191013-install-addons-via-kubeadm.md

```diff
@@ -0,0 +1,450 @@
+
+Internally, the addon-installer library execs `kubectl -k`.
+This introduces a binary and PATH dependency for `kubeadm` when using this feature.
+Furthermore, certain sensible usage of this feature may depend on `git` by proxy of kustomize.
+While `kubectl` may commonly be available in the cluster-installation environment, `git` may not be available.
+ex: The current kinder images do not have `git`.
+
+It may be more appropriate to exec `kustomize` directly or vendor it in for manifest generation so that dryRun and validation of the kustomize addons will work in the absence of a working kube-apiserver. (This is important before `kubeadm init`)
+
+The kubeconfig may be passed to the library as a filepath.
+This works for kubeadm and means less work when exec'ing `kubectl` for the apply logic.
+Should we also support a formal internal type for this?
+
+A more mature addon-installer library may:
+- Vendor kustomize for manifest generation.
+- Use an apiClient type for applying the manifests.
+  ( This may be compatible with some shareable client type supporting dryRun? )
+- Have an go-native git clone implementation
+
+Kustomize currently supports constructing manifests from local directories and git.
+OCI image support in kustomize is currently unimplemented.
+A POC of some of this mechanism is implemented here: https://github.com/ecordell/kpg
+Implementation of this feature could be delegated to another KEP.
+
+### Risks and Mitigations
+
+<!-- What are the risks of this proposal and how do we mitigate.
+Think broadly.
+For example, consider both security and how this will impact the larger kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject. -->
+
+The core of this proposal is that it expands on the curret packaging of addons within the kubeadm binary.
+This binary is currently signed, released, and distributed by kubernetes release machinery.
+
+Since this proposal necessitiates new artifacts and distribution channels for default/"core" components, SIG Release will ultimiately be interested in providing methods of verification and integrity of those artifacts for users.
+
+Git is a strong tool for versioning manifests and patches but may be questionable for asserting source provenance without instrumenting non-default behavior.
+This attack surface may become more evident when mirroring in an airgap environment.
+When fetching/cloning from GitHub the transport will either be verified by HTTPS or an SSH host key.
+
+Distributing kustomize dirs via signed OCI images may be a solution for this.
+
+Proposing additional support for raw manifest lists (with or without kustomization) may also be effective for certain use-cases when sourced over HTTPS from cloud-storage such as GitHub-raw, S3, GCS, or an internal solution like NGINX or Minio.
```

Comment:
We might want to consider some form of signing stuff here. The prospect of fetching a raw manifest from some bucket somewhere (albeit the HTTPS) and applying it over the cluster with an administrative account is kind of scary to me.